### ISSUE TITLE
feat(gp): warped latent-factor regression (conjugacy-preserving output warp)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#196]
## Summary

Let `LatentFactorGPPrior` model the **warped** observations as the linear factor model, so that skewed / heavy-tailed / positive output channels can be handled without breaking the analytic decoder marginalization.

The warp is any `flowjax.bijections.AbstractBijection` with event shape `(P,)` — one marginal transform per output channel. `warp=None` stays the default, so the unwarped path from [A]–[D] keeps no flow dependency.

## Motivation

[A] rests on the observation that every output channel shares the same covariance `C = ZZ^T + sigma^2 I`, which requires Gaussian, homoscedastic, centered outputs. Real high-dimensional output channels — spectra, fluxes, concentrations — are positive, skewed, and differ in scale across channels.

The usual escapes both destroy the method:

- A non-Gaussian likelihood kills the conjugacy the whole collapse depends on.
- Per-output noise `sigma_j^2` breaks the shared-`C` identity that makes the cost `O(NQP)` instead of `O(P^3 N^3)`.

Warping the observations does neither.

## Math specification

Model the warped observations as the linear factor model:

$$
G_\phi^{-1}(y_j) = Z w_j + \varepsilon_j,
\qquad w_j \sim \mathcal{N}(0, I_Q),
\quad \varepsilon_j \sim \mathcal{N}(0, \sigma^2 I_N)
$$

Change of variables gives

$$
\log p(Y \mid Z, \sigma, \phi)
= \underbrace{\log p_{\mathrm{collapsed}}\big(G_\phi^{-1}(Y),\, Z,\, \sigma^2\big)}_{\text{issue [A], unmodified}}
\;+\; \sum_{n=1}^{N}\sum_{j=1}^{P} \log\left|\frac{\partial G_\phi^{-1}}{\partial y}(y_{nj})\right|
$$

**Why the conjugacy survives — the load-bearing point.** The Jacobian term is a function of `phi` and `Y` alone. It is constant with respect to `W`, so marginalizing `W` proceeds exactly as in [A], on the transformed data `Y~ = G^-1(Y)`. It contributes nothing to the gradient w.r.t. `Z` or `sigma`. Therefore:

- the Woodbury structure is untouched — still one `Q x Q` Cholesky,
- the cost stays `O(NQ^2 + NQP + Q^3)` plus one `O(NP)` elementwise pass,
- the decoder posterior is the same matrix normal, read in the warped space:

$$
p(W \mid Z, \tilde Y, \sigma) = \mathcal{MN}_{Q\times P}\big(\sigma^{-2}\Psi^{-1}Z^\top \tilde Y,\; \Psi^{-1},\; I_P\big)
$$

All four parameter blocks — `Z`, kernel hyperparameters, `sigma`, and the warp `phi` — are fitted jointly by gradient descent. Tempering (issue [C]) is unaffected.

## ⚠️ Warp direction is a performance cliff — read before choosing

This likelihood evaluates `G^-1` (flowjax `inverse`: data → base) `N x P` times per gradient step. Some bijections are cheap forward and expensive backward.

*Measured at `N=9, P=5000`:*

| Warp | `transform` | `inverse` | Ratio |
|---|---|---|---|
| `MixtureGaussianCDF` | 40.4 ms | **1624.3 ms** | **40x** |
| `RQSplineMarginal` | 3.3 ms | 3.6 ms | 1.1x |

`MixtureGaussianCDF.inverse` runs a per-dimension **bisection solver** to invert the mixture CDF (its own docstring says so). End-to-end warped log-likelihood at `P = 5000`:

| Warp | Time |
|---|---|
| `MixtureGaussianCDF` | **1642.71 ms** |
| `Invert(MixtureGaussianCDF)` | 35.52 ms (46x faster) |
| `RQSplineMarginal` | **3.58 ms** (459x faster) |

**Default to `RQSplineMarginal`.** Closed-form both directions, and it is the *exact identity at initialization*, so a warped fit starts precisely at the plain GPLFR fit. If a mixture-CDF warp is wanted, wrap it in `flowjax.bijections.Invert` so the closed-form direction is the one used.

> **Note for anyone who also reads the TGP report/issues:** that work recommends `MixtureGaussianCDF` and warns *against* splines, because the TGP ELBO integrates the warp by Gauss-Hermite and piecewise warps destroy its spectral convergence. Neither concern applies here — **the collapsed likelihood is exact, there is no quadrature in this training loop at all.** Quadrature appears only in the post-hoc predictive step. The recommendations differ for principled reasons; do not "harmonise" them.

## Files to create / change

| File | Change |
|---|---|
| `packages/pyrox-gp/src/pyrox_gp/_latent_factor.py` | add `warped_lfr_log_prob`, `warped_decoder_posterior`, `warp_to_base` |
| `packages/pyrox-gp/src/pyrox_gp/_latent_factor_models.py` | `LatentFactorGPPrior` gains `warp: AbstractBijection \| None = None`; `condition` and `predict` route through it |
| `packages/pyrox-gp/pyproject.toml` | `flowjax>=0.9` as a dependency; `[flows]` extra for `gauss-flows` |
| `packages/pyrox-gp/src/pyrox_gp/__init__.py` | export the new names |
| `packages/pyrox-gp/tests/gp/test_warped_latent_factor.py` | **new** |
| `docs/api/gp.md` | add the new entries |

## Dependencies

```toml
dependencies = [
    # ... existing ...
    "flowjax>=0.9",     # AbstractBijection protocol + Invert/Chain/RationalQuadraticSpline.
                        # Shares lineax/equinox/jax/jaxtyping with pyrox-gp already;
                        # net new transitive deps are just paramax and tqdm.
]

[project.optional-dependencies]
# RQSplineMarginal (the recommended warp, event shape (P,)),
# gaussian_total_correlation, and the closed-form RBIG warm start.
# Pulls scikit-learn / diffrax / matfree, hence an extra rather than a hard dep.
flows = ["gauss-flows @ git+https://github.com/jejjohnson/gauss_flows.git@<pin>"]
```

Pin `gauss-flows` by commit, matching the existing `gaussx` / `geonnax` convention.

## Reference implementation

Validated — see "Validation evidence". Docstrings are Google style with `$...$` math; **no Sphinx/RST markup** (`test_docstrings.py` rejects it).

```python
def warp_to_base(
    warp: AbstractBijection,
    Y: Float[Array, "N P"],
) -> tuple[Float[Array, "N P"], Float[Array, ""]]:
    """Map observations into the space where the factor model is linear.

    The warp has event shape ``(P,)`` -- one marginal transform per output
    channel -- and is applied independently to each of the ``N`` rows.

    Args:
        warp: Bijection with event shape ``(P,)``. flowjax convention:
            ``transform`` maps base to data, ``inverse`` maps data to base,
            so this uses ``inverse``.
        Y: Observations of shape ``(N, P)``.

    Returns:
        Tuple of ``(Y_tilde, total_log_det)``.
    """
    if warp.cond_shape is not None:
        raise ValueError(
            "Conditional warps are not supported here: the log-det Jacobian "
            "would depend on the inputs, and the per-channel marginal "
            "interpretation no longer holds."
        )
    Ytil, log_det = jax.vmap(warp.inverse_and_log_det)(Y)
    return Ytil, jnp.sum(log_det)


def warped_lfr_log_prob(Y, Z, noise_var, warp, *, jitter=1e-12):
    r"""Collapsed latent-factor log-likelihood on warped observations.

    Models $G^{-1}(y_j) = Z w_j + \epsilon$, so

    $$
    \log p(Y) = \log p_{\mathrm{collapsed}}(G^{-1}(Y), Z, \sigma^2)
              + \sum_{n,j} \log \left| \partial G^{-1} / \partial y \right|
    $$

    The Jacobian term is free of $Z$, $W$ and $\sigma$, so the analytic
    decoder marginalization of
    [`collapsed_lfr_log_prob`][pyrox_gp.collapsed_lfr_log_prob] is
    unchanged and the cost stays linear in $P$.

    !!! warning "Warp direction is a performance cliff"
        This evaluates ``G^{-1}`` (``inverse``) on every step.
        ``MixtureGaussianCDF.inverse`` runs a bisection solver and is ~40x
        its forward cost -- 459x slower end-to-end than the alternative.
        Prefer ``gauss_flows.RQSplineMarginal`` (closed-form both ways, and
        the exact identity at initialization), or wrap a mixture-CDF warp
        in ``flowjax.bijections.Invert``.
    """
    Ytil, log_det = warp_to_base(warp, Y)
    return collapsed_lfr_log_prob(Ytil, Z, noise_var, jitter=jitter) + log_det


def warped_decoder_posterior(Y, Z, noise_var, warp, *, jitter=1e-12):
    """Matrix-normal decoder posterior, read in the warped space.

    Identical to [`decoder_posterior`][pyrox_gp.decoder_posterior] applied
    to ``G^{-1}(Y)`` -- the warp does not change the conjugacy.
    """
    Ytil, _ = warp_to_base(warp, Y)
    return decoder_posterior(Ytil, Z, noise_var, jitter=jitter)
```

Model surface:

```python
class LatentFactorGPPrior(eqx.Module):
    kernels: tuple[Kernel, ...]
    X: Float[Array, "N D"]
    warp: AbstractBijection | None = None
    latent_noise: float = 1e-3
    jitter: float = 1e-6
```

Prediction is two-stage: the factor-model predictive is Gaussian in the warped space (the exact moment formula from [A]), then pushed through `G` (`transform`) by Gauss-Hermite quadrature to get observation-space moments. **Here** the smoothness caveat from the TGP work does apply — but only to prediction, not training, so a moderate order (32) is fine. Return `E[G(f)]`, never `G(E[f])`.

## Usage

```python
from gauss_flows import RQSplineMarginal
from pyrox_gp import RBF, LatentFactorGPPrior

prior = LatentFactorGPPrior(
    kernels=tuple(RBF(pyrox_name=f"RBF_q{q}", input_dim=X.shape[1]) for q in range(Q)),
    X=X,
    warp=RQSplineMarginal(n_bins=8, interval=4.0, shape=(Y.shape[1],)),
)
# identity at init => step 0 is exactly the unwarped GPLFR fit
```

## Acceptance criteria

New file `packages/pyrox-gp/tests/gp/test_warped_latent_factor.py`, `jax.config.update("jax_enable_x64", True)` at module top.

- [ ] **The warped density is a proper density.** With `N = P = Q = 1`, numerically integrate `exp(warped_lfr_log_prob(y, ...))` over a wide grid; assert it equals `1` to `1e-6`. **Also assert it does *not* equal 1 when the log-det term is omitted** — this is what pins the Jacobian as load-bearing rather than cosmetic.
- [ ] **Identity-init warp reduces exactly to [A].** `RQSplineMarginal(n_bins=8, interval=4.0, shape=(P,))` at construction is the exact identity, so `warped_lfr_log_prob(Y, Z, s2, warp)` must equal `collapsed_lfr_log_prob(Y, Z, s2)` **exactly** (tolerance `0.0`, not a tolerance band).
- [ ] **Conjugacy survives.** `warped_decoder_posterior` matches a per-column Bayesian linear regression computed on `G^-1(Y)` by hand, to `1e-10`, for both mean and covariance.
- [ ] **Gradients reach all four blocks** — `Z`, `noise_var`, kernel hyperparameters, and the warp leaves — and are finite.
- [ ] **Cost stays linear in `P`.** Time at `P = 10^3, 10^4, 5x10^4` with `RQSplineMarginal`; assert sub-quadratic growth. Mark `@pytest.mark.slow`.
- [ ] **Direction-trap regression.** Assert `MixtureGaussianCDF` and `Invert(MixtureGaussianCDF)` give the *same* log-likelihood up to the bijection's inversion tolerance, so the recommended workaround is provably equivalent and not a silent change of model. (Do not assert on timings in CI — machine-dependent.)
- [ ] **`warp=None` requires no flow import.** A test that the unwarped path works with `flowjax` absent, or at minimum that `pyrox_gp` imports without `gauss_flows` installed.
- [ ] **Conditional warps are rejected** with the clear `ValueError` above.
- [ ] **Shape validation.** A warp whose event shape is not `(P,)` raises a `ValueError` naming the expected shape.
- [ ] **Predictive moments** vs Monte Carlo in observation space, `2%` relative. Mark `@pytest.mark.slow`.
- [ ] Assert `E[G(f)] != G(E[f])` for a non-identity warp, so nobody "simplifies" the predictive later.

## Validation evidence

Executed against the reference implementation:

| Check | Result |
|---|---|
| Warped density integral, `N=P=Q=1`, 400k-point grid | `1.00000000` |
| Same, log-det term removed | `0.98487747` |
| Decoder posterior on warped data vs brute force | mean `2.7e-13`, cov `3.6e-13` |
| Identity-init warp vs plain collapsed likelihood | `diff = 0.00e+00` |
| Gradients into `Z`, `sigma`, warp | all finite |
| `RQSplineMarginal` cost, `P = 10^3 / 10^4 / 5x10^4` | `0.81 / 8.60 / 52.83 ms` |
| Direction trap, `P = 5000` | `1642.71` / `35.52` / `3.58 ms` |

## Caveats to document in the class docstring

- **`Y` must sit inside the warp's support.** `RQSplineMarginal` is linear outside `[-interval, interval]`; scale `Y` sensibly or the warp degenerates to affine.
- **Identifiability gets worse.** A per-channel affine warp is degenerate with `sigma` and with the columns of `W`. Identity init plus a distinct learning-rate group (issue [D]) is the mitigation.
- **The warp is fitted to the marginals of `Y`, not to the residuals.** A warp that Gaussianizes raw channel marginals is not necessarily the one that Gaussianizes the noise. A genuine modelling approximation, not a bug.
- **Invertibility is mandatory here**, unlike in the TGP case where a bound holds for any monotone link. The change of variables requires a bijection.

## Out of scope

- Input-dependent / conditional warps (rejected above). The Jacobian would still be `Z`-free so the conjugacy argument survives, but the per-channel marginal reading does not, and it interacts with the tempering. Separate investigation.
- Warping the **latents** `Z = G(F)`. Also conjugacy-preserving, but it only reshapes the effective prior on an already point-estimated quantity — much lower value. Note it and move on.
- A flow-based (non-linear) decoder. Destroys the collapse. Not a follow-up.
- `fit_rbig` / `MixtureGaussianCDF.from_data` warm starts. Worth having later; the identity init is sufficient to land this.

## Before reporting done

```bash
uv run pytest -q
uv run --group lint ruff check .
uv run --group lint ruff format --check .
uv run --group typecheck ty check packages/pyrox-gp/src/pyrox_gp
```

**Comment by jejjohnson** [comment]
Closing as a duplicate of #201 — the two bodies are word-for-word identical (same math spec, reference implementation, acceptance criteria, and validation evidence). Work will proceed on #201, which sits with the rest of the lettered series (#199/#200/#202–#204).
